osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs

```diff
@@ -540,18 +537,21 @@ private void updatePlaylist(MultiplayerRoomSettings settings, APIBeatmapSet onli
             var mods = settings.RequiredMods.Select(m => m.ToMod(ruleset));
             var allowedMods = settings.AllowedMods.Select(m => m.ToMod(ruleset));
 
-            PlaylistItem playlistItem = new PlaylistItem
-            {
-                ID = playlistItemId,
-                Beatmap = { Value = beatmap },
-                Ruleset = { Value = ruleset.RulesetInfo },
-            };
+            // Update an existing playlist item from the API room, or create a new item.
+            var playlistItem = apiRoom.Playlist.FirstOrDefault(i => i.ID == settings.PlaylistItemId) ?? new PlaylistItem();
 
+            playlistItem.ID = settings.PlaylistItemId;
+            playlistItem.Beatmap.Value = beatmap;
+            playlistItem.Ruleset.Value = ruleset.RulesetInfo;
+            playlistItem.RequiredMods.Clear();
             playlistItem.RequiredMods.AddRange(mods);
+            playlistItem.AllowedMods.Clear();
             playlistItem.AllowedMods.AddRange(allowedMods);
 
-            apiRoom.Playlist.Clear(); // Clearing should be unnecessary, but here for sanity.
-            apiRoom.Playlist.Add(playlistItem);
+            if (!apiRoom.Playlist.Contains(playlistItem))
```

Comment:
This is hinging on the fact that `PlaylistItem` doesn't implement an `Equals(object obj)` override, so it's checking based on reference. Not sure how I feel about this.

Might be best to just split out the two cases (for new item vs updating item) into an if-else?

Reply:
Sure. I'm not going to dwell on this too much since it's not a huge change. But `PlaylistItem` does implement `IEquatable` (not sure why you're referencing the `object` variety).

Reply:
To my knowledge `List.Contains` won't use the IEquatable implementation without the object override being implemented, which is what is allowing this to work correctly? Because if it *was* using the custom implementation it would not be working - that implementation compares more than just the ID, which would cause the contains to always be `false` here).